File: fsharp/dmmf/domain/GizmoCode.fs
namespace Domain

open System.Text.RegularExpressions

type GizmoCode = private GizmoCode of string
module GizmoCode = 
    let create code =
        if Regex.IsMatch(code, @"G\d{3}") then
            Ok (GizmoCode code)
        else
            Error "`GizmoCode` must begin with a 'G' and be followed by 3 digits"

    let value (GizmoCode code) = code

File: rust/src/domain/gizmo_code.rs
use regex::Regex;

#[derive(PartialEq, Debug)]
pub struct GizmoCode(String);

impl GizmoCode {
    pub fn create(code: &str) -> Result<GizmoCode, String> {
        let re = Regex::new(r"G\d{3}").expect(r"G\d{3} is a valid regex");
        if re.is_match(code) {
            Ok(GizmoCode(String::from(code)))
        } else {
            Err(String::from(
                "`GizmoCode` must begin with a 'G' and be followed by 3 digits",
            ))
        }
    }

    pub fn value(&self) -> &str {
        &self.0
    }
}
